### ISSUE TITLE
Suggestion for new hook feature: When certificate request is succeeded execute hook

### ISSUE THREAD

**Issue by flphrrr** [grindsa/acme2certifier#87]
Maybe in some cases, one can make use of the hook options on [acme.sh](https://github.com/acmesh-official/acme.sh) client, but when some restricted actions must be executed on a controlled zone, like creation of an object in LDAP or AD by specific service accounts, it would be nice to consider the possibility to execute a hooks once the certificate request has been succeeded.

On our case when creating a Certificate request using AD CA, is required to ensure that the object exists in Active Directory due specific authorisation mechanisms, could it be a feasible option to have this hook, accepting some parameters like the fqdn?

We would be happy to create a PR for this new feature

**Comment by grindsa** [comment]
PRs are welcome at any time. No need to ask for it. Do you see this as a feature of the ca-handler or rather as general hook to be executed after successful certificate enrolment?

**Comment by flphrrr** [comment]
My understanding is that could be a general one, independently of the ca-handler used, if we attach this feature to an specific ca-handler we are at risk to require to modify each handler to make it possible, while adding a general hook, it could maybe cover more use cases, with a minimum code change, requiring only to enable the hook based on the specific script provided..  As usually hooks could be dangerous and provide slowness on the process, I would suggest to avoid to have error control on the results of the hook on acme_srv and simply execute it without considering the results of the hook, is up to the user to determine a proper hook scripting.. what do you think?

**Comment by grindsa** [comment]
Sounds good. Let me throw some rough ideas coming into my mind:

- we should broaden the scope and allow the definition of pre- and/or post-enrolment hooks
- the hooks should be called from [acme_srv/certificate.py _enroll_and_store() function](https://github.com/grindsa/acme2certifier/blob/devel/acme_srv/certificate.py#L221)

- the hooks can be loaded from a file to be specified in the `[Certificate]` section of acme_srv.cfg via pre-defined function calls so something like 
```
def preinstallation_hook(certificate_name, order_name, csr):
  """ preinstallation hook """
  return True

def preinstallation_hook(certificate_name, order_name, csr, certificate, certificate_raw, poll_identifier):
  """ postinstallation hook """
  return True
````
- the function can return `True/False` to allow monitoring of the hook execution. I do not see performance issues here as `_enroll_and_store()` already runs in a separate thread to ensure client responses in a timely manner 

Thoughts?

**Comment by flphrrr** [comment]
For me looks perfect, I was thinking on how to be able to pass specific parameter to the hook, but the definition examples provided make it more clear.

For me looks great.. we´ll start checking it.. 👍

**Comment by grindsa** [comment]
Included in v0.21
